Sources/MapboxDirections/RouteLeg.swift

```diff
@@ -98,18 +82,24 @@ open class RouteLeg: Codable {
         try container.encode(profileIdentifier, forKey: .profileIdentifier)
         
         if segmentDistances != nil || expectedSegmentTravelTimes != nil || segmentSpeeds != nil || segmentCongestionLevels != nil || segmentMaximumSpeedLimits != nil {
-            var annotationContainer = container.nestedContainer(keyedBy: AnnotationCodingKeys.self, forKey: .annotation)
-            try annotationContainer.encodeIfPresent(segmentDistances, forKey: .segmentDistances)
-            try annotationContainer.encodeIfPresent(expectedSegmentTravelTimes, forKey: .expectedSegmentTravelTimes)
-            try annotationContainer.encodeIfPresent(segmentSpeeds, forKey: .segmentSpeeds)
-            try annotationContainer.encodeIfPresent(segmentCongestionLevels, forKey: .segmentCongestionLevels)
             
-            if let speedLimitDescriptors = segmentMaximumSpeedLimits?.map({ SpeedLimitDescriptor(speed: $0) }) {
-                try annotationContainer.encode(speedLimitDescriptors, forKey: .segmentMaximumSpeedLimits)
-            }
+            let annotation = RouteLegAnnotation(segmentDistances: segmentDistances,
+                                                expectedSegmentTravelTimes: expectedSegmentTravelTimes,
+                                                segmentSpeeds: segmentSpeeds,
+                                                segmentCongestionLevels: segmentCongestionLevels,
+                                                segmentMaximumSpeedLimits: segmentMaximumSpeedLimits)
+            try container.encode(annotation, forKey: .annotation)
         }
     }
 
+    func updateAnnotationData(from annotation: RouteLegAnnotation?) {
+        segmentDistances = annotation?.segmentDistances
+        expectedSegmentTravelTimes = annotation?.expectedSegmentTravelTimes
+        segmentSpeeds = annotation?.segmentSpeeds
+        segmentCongestionLevels = annotation?.segmentCongestionLevels
+        segmentMaximumSpeedLimits = annotation?.segmentMaximumSpeedLimits
```

Comment:
I think we can live with this repetitiveness for now, but long term I’m concerned about potentially adding more attribute options and forgetting to copy one of them over here. To avoid that problem, and also to clean up the RouteLeg interface a bit, we could make RouteLegAnnotation public and replace the various segment* properties on RouteLeg with a single `attributes` property that we’d simply replace when refreshing.

We can take care of this refactoring as tail work in a future release, preserving backwards compatibility by leaving the segment* properties as computed properties.

Reply:
I moved this copying to a more obvious location in the setter of the `attributes` property, taking advantage of copy-by-value for structs.